lib/http_negotiate.c

```diff
@@ -101,38 +104,76 @@ CURLcode Curl_input_negotiate(struct connectdata *conn, bool proxy,
 
 CURLcode Curl_output_negotiate(struct connectdata *conn, bool proxy)
 {
-  struct negotiatedata *neg_ctx = proxy ? &conn->data->state.proxyneg :
-    &conn->data->state.negotiate;
+  struct negotiatedata *neg_ctx = proxy ? &conn->proxyneg :
+    &conn->negotiate;
+  struct auth *authp = proxy ? &conn->data->state.authproxy :
+    &conn->data->state.authhost;
   char *base64 = NULL;
   size_t len = 0;
   char *userp;
   CURLcode result;
 
-  result = Curl_auth_create_spnego_message(conn->data, neg_ctx, &base64, &len);
-  if(result)
-    return result;
+  authp->done = FALSE;
 
-  userp = aprintf("%sAuthorization: Negotiate %s\r\n", proxy ? "Proxy-" : "",
-                  base64);
+  if(neg_ctx->state != GSS_AUTHDONE && neg_ctx->state != GSS_AUTHSUCC) {
+
+    if(!neg_ctx->context) {
+      result = Curl_input_negotiate(conn, proxy, "Negotiate");
+      if(result)
+        return result;
+    }
+
+    result = Curl_auth_create_spnego_message(conn->data,
+      neg_ctx, &base64, &len);
+    if(result)
+      return result;
+
+    userp = aprintf("%sAuthorization: Negotiate %s\r\n", proxy ? "Proxy-" : "",
+      base64);
+
+    if(proxy) {
+      Curl_safefree(conn->allocptr.proxyuserpwd);
+      conn->allocptr.proxyuserpwd = userp;
+    }
+    else {
+      Curl_safefree(conn->allocptr.userpwd);
+      conn->allocptr.userpwd = userp;
+    }
+
+    free(base64);
+
+    if(userp == NULL) {
+      return CURLE_OUT_OF_MEMORY;
+    }
+
+    neg_ctx->state = GSS_AUTHSENT;
+  #ifdef HAVE_GSSAPI
+    if(neg_ctx->status == GSS_S_COMPLETE || neg_ctx->status == GSS_S_CONTINUE_NEEDED) {
+      neg_ctx->state = GSS_AUTHDONE;
```

Comment:
How can this be done when the context says continue needed?

Reply:
When `GSS_S_CONTINUE_NEEDED` is returned it can happen that the server already accepts the authentication when the negotiate data returned from this function is sent to the server. So the request body needs to be sent to the server also on `GSS_S_CONTINUE_NEEDED`, otherwise the server may successfully process the request without the request body.

See Google Chrome source code:

https://cs.chromium.org/chromium/src/net/http/http_auth_gssapi_posix.cc?type=cs&q=GSS_S_CONTINUE_NEEDED&sq=package:chromium&l=768

https://cs.chromium.org/chromium/src/net/http/http_auth_sspi_win.cc?type=cs&q=SEC_I_CONTINUE_NEEDED&sq=package:chromium&l=112

Both return codes are accepted as "OK" and not differentiated.

Reply:
That's still no right because you have to complete the context loop first. Negotiate is a client-first mechanism. Unless you init_sec_context is completed, you shall not trust the peer.

Reply:
Thank you for your answere!

What are your recommended changes to fix this?

How can I detect that the FINAL/LAST request to the server has to be issued and the request-body (instead of `Content-Length: 0` while the negotiate phase) can be sent to the server?
When only performing a GET request, it also works setting GSS_AUTHDONE only on SEC_E_OK.

The most important thing for me is to get this work on Windows with SSPI.
My problem is that when `InitializeSecurityContext` returns `SEC_I_CONTINUE_NEEDED` and the token from that call is sent to the server, the server already accepts it in some cases, processes the request, and does not send HTTP 401 any more. And when the request body is not sent in this request, the POST request processes incorrectly because the body is missing and `Content-Length`is still 0 as cURL thinks it is still in negotiate phase.